### PR TITLE
[UNR-5638] Add better logging for the RegisterAutoDestroyActor tests

### PR DESCRIPTION
**Pull request by mironec**
#### Description
Some better logging, small refactors.
We should get more information if this fails.

#### Release note
Not required.

#### Tests
Ran the tests 300 times.

#### Documentation
N/A

#### Primary reviewers
idk

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp [hunk 1]
```diff
@@ -50,44 +50,40 @@ void ARegisterAutoDestroyActorsTestPart1::PrepareTest()
...
+									 TEXT("Servers should observe the correct number of characters in their world.")))
 				{
-					if (It->HasAuthority())
-					{
-						AssertTrue(IsValid(*It), FString::Printf(TEXT("Registering ACharacter for destruction: %s"), *GetNameSafe(*It)));
-						RegisterAutoDestroyActor(*It);
-					}
+					AssertTrue(IsValid(FoundAuthoritativeCharacter), FString::Printf(TEXT("Registering ACharacter for destruction: %s"),
+																					 *GetNameSafe(FoundAuthoritativeCharacter)));
+					RegisterAutoDestroyActor(FoundAuthoritativeCharacter);
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:81]: Why are we only registering the last FoundAuthorativeCharacter for destroying?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp [hunk 2]
```diff
@@ -50,44 +50,40 @@ void ARegisterAutoDestroyActorsTestPart1::PrepareTest()
...
 
 	{ // Step 3 - Destroy by all servers that have authority
 		AddStep(
-			TEXT("SERVER_ALL_RegisterAutoDestroyActors"), FWorkerDefinition::AllServers,
-			[this]() -> bool {
+			TEXT("SERVER_ALL_RegisterAutoDestroyActors"), FWorkerDefinition::AllServers, nullptr, nullptr,
+			[this](float DeltaTime) {
 				int NumCharactersFound = 0;
-				int NumCharactersExpected = 1;
+				int NumCharactersExpected = 1; // 1, because we filter by authority
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:65]: minor: `uint32`?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp [hunk 3]
```diff
@@ -106,11 +102,16 @@ void ARegisterAutoDestroyActorsTestPart2::PrepareTest()
...
 		StepDefinition.TimeLimit = 0.0f;
 		StepDefinition.NativeStartEvent.BindLambda([this]() {
+			int NumCharactersFound = 0;
+			int NumCharactersExpected = 0;
 			UWorld* World = GetWorld();
-			TActorIterator<ACharacter> It(World);
-			bool bHasCharacter = static_cast<bool>(It);
-			AssertFalse(bHasCharacter, FString::Printf(TEXT("Cleanup of ACharacter successful, no ACharacter found by %s"),
-													   *GetLocalFlowController()->GetDisplayName()));
+			for (TActorIterator<ACharacter> It(World); It; ++It)
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:108]: Minor could use: `TActorRange` for slightly nicer ranged for semantics.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp (hunk 1 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:81]: there should only be one, which we also check later

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp (hunk 3 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:108]: Ah, that's nice, we should probably use that more often

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp (hunk 2 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:65]: hmm, I don't know if we actually have a preference here, we seem to use just `int` in quite a few places

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp (hunk 1 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:81]: or I guess earlier

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp [hunk 4]
```diff
@@ -42,52 +42,48 @@ void ARegisterAutoDestroyActorsTestPart1::PrepareTest()
...
+						FoundAuthoritativeCharacter = Character;
 					}
 				}
-
-				return NumCharactersFound == NumCharactersExpected;
-			},
-			[this]() {
-				UWorld* World = GetWorld();
-				for (TActorIterator<ACharacter> It(World); It; ++It)
+				if (RequireEqual_Int(NumCharactersFound, NumCharactersExpected,
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:76]: Very minor: If you kept a vector of `FoundAuthoritativeCharacter `and check its length instead I think it communicates its intent a bit better. You now have strong implicit link between `NumCharactersExpected `and how you store `FoundAuthoritativeCharacter`.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp (hunk 1 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:81]: Got it.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp [hunk 5]
```diff
@@ -42,52 +42,48 @@ void ARegisterAutoDestroyActorsTestPart1::PrepareTest()
 		AddStep(
 			TEXT("CLIENT_ALL_CheckActorsSpawned"), FWorkerDefinition::AllClients, nullptr, nullptr,
 			[this](float DeltaTime) {
-				int NumCharactersFound = 0;
-				int NumCharactersExpected = GetNumberOfServerWorkers();
+				int32 NumCharactersFound = 0;
+				int32 NumCharactersExpected = GetNumberOfServerWorkers();
 				UWorld* World = GetWorld();
-				for (TActorIterator<ACharacter> It(World); It; ++It)
+				for (ACharacter* Character : TActorRange<ACharacter>(World))
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:48]: Very very minor: but unreal has some stl like algo's, in this case the mildly awkward CountIf will work. (No unconditional count is available sadly).
```
const int32 NumCharactersFound = Algo::CountIf(TActorRange<ACharacter>(World), [](){ return true;}); 
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp [hunk 6]
```diff
@@ -42,52 +42,48 @@ void ARegisterAutoDestroyActorsTestPart1::PrepareTest()
...
-				int NumCharactersFound = 0;
-				int NumCharactersExpected = 1;
+			TEXT("SERVER_ALL_RegisterAutoDestroyActors"), FWorkerDefinition::AllServers, nullptr, nullptr,
+			[this](float DeltaTime) {
+				int32 NumCharactersFound = 0;
+				int32 NumCharactersExpected = 1; // 1, because we filter by authority
+				ACharacter* FoundAuthoritativeCharacter = nullptr;
 				UWorld* World = GetWorld();
-				for (TActorIterator<ACharacter> It(World, ACharacter::StaticClass()); It; ++It)
+				for (ACharacter* Character : TActorRange<ACharacter>(World))
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:68]: Very very minor: more algo's
```
TArray<ACharacter*> FoundAuthoritativeCharacters;
Algo::ForEachIf(TActorRange<ACharacter>(World), [](ACharacter* Character)
{
	return Character->HasAuthority();
},
[&FoundAuthoritativeCharacters](ACharacter* Character)
{
	FoundAuthoritativeCharacters.Push(Character);
});

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp [hunk 7]
```diff
@@ -42,52 +42,48 @@ void ARegisterAutoDestroyActorsTestPart1::PrepareTest()
 		AddStep(
 			TEXT("CLIENT_ALL_CheckActorsSpawned"), FWorkerDefinition::AllClients, nullptr, nullptr,
 			[this](float DeltaTime) {
-				int NumCharactersFound = 0;
-				int NumCharactersExpected = GetNumberOfServerWorkers();
+				int32 NumCharactersFound = 0;
+				int32 NumCharactersExpected = GetNumberOfServerWorkers();
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:46]: nit - should be const

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp [hunk 8]
```diff
@@ -18,14 +18,14 @@ void ARegisterAutoDestroyActorsTestPart1::PrepareTest()
 	{ // Step 1 - Spawn Actor On Auth
 		AddStep(TEXT("SERVER_1_Spawn"), FWorkerDefinition::Server(1), nullptr, [this]() {
 			UWorld* World = GetWorld();
-			int NumVirtualWorkers = GetNumberOfServerWorkers();
+			int32 NumVirtualWorkers = GetNumberOfServerWorkers();
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:21]: nit - should be const

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp [hunk 9]
```diff
@@ -18,14 +18,14 @@ void ARegisterAutoDestroyActorsTestPart1::PrepareTest()
...
 
 			// spawn 1 per server worker
 			// since the information about positioning of the virtual workers is currently hidden, will assume they are all around zero
 			// and spawn them in that radius
 			FVector SpawnPosition = FVector(200.0f, -200.0f, 0.0f);
 			FRotator SpawnPositionRotator = FRotator(0.0f, 360.0f / NumVirtualWorkers, 0.0f);
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:27]: nit - hopefully NumVirtualWorkers can never be zero zero.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp [hunk 10]
```diff
@@ -18,14 +18,14 @@ void ARegisterAutoDestroyActorsTestPart1::PrepareTest()
 	{ // Step 1 - Spawn Actor On Auth
 		AddStep(TEXT("SERVER_1_Spawn"), FWorkerDefinition::Server(1), nullptr, [this]() {
 			UWorld* World = GetWorld();
-			int NumVirtualWorkers = GetNumberOfServerWorkers();
+			int32 NumVirtualWorkers = GetNumberOfServerWorkers();
 
 			// spawn 1 per server worker
 			// since the information about positioning of the virtual workers is currently hidden, will assume they are all around zero
 			// and spawn them in that radius
 			FVector SpawnPosition = FVector(200.0f, -200.0f, 0.0f);
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:26]: nit - what do these magic numbers mean?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp [hunk 11]
```diff
@@ -18,14 +18,14 @@ void ARegisterAutoDestroyActorsTestPart1::PrepareTest()
...
 			FRotator SpawnPositionRotator = FRotator(0.0f, 360.0f / NumVirtualWorkers, 0.0f);
-			for (int i = 0; i != NumVirtualWorkers; ++i)
+			for (int32 i = 0; i != NumVirtualWorkers; ++i)
 			{
 				ACharacter* Character = World->SpawnActor<ACharacter>(SpawnPosition, FRotator::ZeroRotator);
 				AssertTrue(IsValid(Character),
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:31]: Would it be advisable to abort the test here as well as log an error?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp [hunk 12]
```diff
@@ -106,11 +102,16 @@ void ARegisterAutoDestroyActorsTestPart2::PrepareTest()
 		StepDefinition.StepName = TEXT("Check No Worker Has Characters");
 		StepDefinition.TimeLimit = 0.0f;
 		StepDefinition.NativeStartEvent.BindLambda([this]() {
+			int32 NumCharactersFound = 0;
+			int32 NumCharactersExpected = 0;
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:106]: nit - should be const

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp [hunk 13]
```diff
@@ -42,52 +42,48 @@ void ARegisterAutoDestroyActorsTestPart1::PrepareTest()
...
 	{ // Step 3 - Destroy by all servers that have authority
 		AddStep(
-			TEXT("SERVER_ALL_RegisterAutoDestroyActors"), FWorkerDefinition::AllServers,
-			[this]() -> bool {
-				int NumCharactersFound = 0;
-				int NumCharactersExpected = 1;
+			TEXT("SERVER_ALL_RegisterAutoDestroyActors"), FWorkerDefinition::AllServers, nullptr, nullptr,
+			[this](float DeltaTime) {
+				int32 NumCharactersFound = 0;
+				int32 NumCharactersExpected = 1; // 1, because we filter by authority
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:65]: nit - const

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp [hunk 14]
```diff
@@ -42,52 +42,48 @@ void ARegisterAutoDestroyActorsTestPart1::PrepareTest()
...
+				ACharacter* FoundAuthoritativeCharacter = nullptr;
 				UWorld* World = GetWorld();
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:67]: Unless you want to check its validity, I don't think this add's much value.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp (hunk 13 above)
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:65]: Useful comment thanks!

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp (hunk 5 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp:48]: Tried it, but unfortunately you also have to take the first argument in the predicate (even if you don't use it), and our auto-formatter spreads it across two lines, so I think the reading complexity is the same as the for-each case

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp]: This gives more const and is a bit more explicit in its intent (it uses a function called Count) but very minor so fine as is.

**Review by martin-improbable — APPROVED** [review]
LGTM, left some optional nits.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp (hunk 10 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/RegisterAutoDestroyActorsTest/RegisterAutoDestroyActorsTest.cpp]: added a comment, but it's basically just for visuals, not too important - the more important part is the rotation, which is explained

**Pull request merged**